pennylane/devices/qubit/simulate.py

```diff
@@ -0,0 +1,102 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Simulate a quantum script."""
+
+# pylint: disable=protected-access
+import pennylane as qml
+from pennylane.wires import Wires
+from pennylane.measurements import StateMeasurement, MeasurementProcess
+from pennylane.typing import TensorLike
+
+from .initialize_state import create_initial_state
+from .apply_operation import apply_operation
+
+
+def measure_state_diagonalizing_gates(
+    measurementprocess: StateMeasurement, state: TensorLike
+) -> TensorLike:
+    """Apply a measurement to state when the measurement process has an observable with diagonalizing gates.
+
+    Args:
+        measurementprocess (StateMeasurement): measurement to apply to the state
+        state (TensorLike): state to apply the measurement to
+
+    Returns:
+        TensorLike: the result of the measurement
+    """
+    total_indices = len(state.shape)
+    wires = Wires(range(total_indices))
+
+    for op in measurementprocess.obs.diagonalizing_gates():
+        state = apply_operation(op, state)
+
+    return measurementprocess.process_state(qml.math.flatten(state), wires)
+
+
+def measure(measurementprocess: MeasurementProcess, state: TensorLike) -> TensorLike:
+    """Apply a measurement process to a state.
+
+    Args:
+        measurementprocess (MeasurementProcess): measurement process to apply to the state
+        state (TensorLike): the state to measure
+
+    Returns:
+        Tensorlike: the result of the measurement
+    """
+    if isinstance(measurementprocess, StateMeasurement):
+        if measurementprocess.obs is None:
+            # no need to apply diagonalizing gates
+            total_indices = len(state.shape)
+            wires = Wires(range(total_indices))
+            return measurementprocess.process_state(qml.math.flatten(state), wires)
+
+        if measurementprocess.obs.has_diagonalizing_gates:
+            return measure_state_diagonalizing_gates(measurementprocess, state)
+
+    raise NotImplementedError
+
+
+def simulate(circuit: qml.tape.QuantumScript) -> tuple:
+    """Simulate a single quantum script.
+
+    This is an internal function that will be called by successor to ``default.qubit``.
+
+    Args:
+        circuit (.QuantumScript): The single circuit to simulate
+
+    Returns:
+        tuple(ndarray): The results of the simulation
+
+    Note that this function can return measurements for non-commuting observables simultaneously.
+
+    It does currently not support sampling or observables without diagonalizing gates.
+
+    This function assumes that all operations provide matrices.
+
+    >>> qs = qml.tape.QuantumScript([qml.RX(1.2, wires=0)], [qml.expval(qml.PauliZ(0)), qml.probs(wires=(0,1))])
+    >>> simulate(qs)
+    (0.36235775447667357,
+    tensor([0.68117888, 0.        , 0.31882112, 0.        ], requires_grad=True))
+
+    """
+    if set(circuit.wires) != set(range(circuit.num_wires)):
```

Comment:
Be careful that `set` orders the wires:
```pycon
>>> {3, 1, 0, 2}
{0, 1, 2, 3}
```

Reply:
This step is to turn wires into dimensions of the state vector. If `circuit.wires` was `(3,1,0,2)`, then each wire still corresponds to a dimension on a (2,2,2,2) state vector. In which case, we can simply leave those wires be.

Reply:
Okay! Just making sure you were aware of this 😄